test_runner/fixtures/zenith_fixtures.py

```diff
@@ -1333,6 +1333,82 @@ def vanilla_pg(test_output_dir: str) -> Iterator[VanillaPostgres]:
         yield vanilla_pg
 
 
+class RemotePostgres(PgProtocol):
+    def __init__(self, pg_bin: PgBin, remote_connstr: str):
+        self.pg_bin = pg_bin
+        # The remote server is assumed to be running already
+        self.running = True
+        self.remote_connstr = remote_connstr
+        self.username = None
+        self.password = None
+        self.dbname = None
+        self.schema = None
+
+    def configure(self, options: List[str]):
+        raise Exception('cannot change configuration of remote Posgres instance')
+
+    def start(self):
+        raise Exception('cannot start a remote Postgres instance')
+
+    def stop(self):
+        raise Exception('cannot stop a remote Postgres instance')
+
+    def get_subdir_size(self, subdir) -> int:
+        # TODO: Could use the server's Generic File Acccess functions if superuser.
+        # See https://www.postgresql.org/docs/14/functions-admin.html#FUNCTIONS-ADMIN-GENFILE
+        raise Exception('cannot get size of a Postgres instance')
+
+    def connstr(self,
+                *,
+                dbname: Optional[str] = None,
+                schema: Optional[str] = None,
+                username: Optional[str] = None,
+                password: Optional[str] = None,
+                statement_timeout_ms: Optional[int] = None) -> str:
+        """
+        Build a libpq connection string for the Postgres instance.
+        """
+
+        username = username or self.username
+        password = password or self.password
+        dbname = dbname or self.dbname or "postgres"
+        schema = schema or self.schema
+        res = f'{self.remote_connstr} dbname={dbname}'
```

Comment:
This is exactly the cause of the Unauthorized error because it generates weird connection string which changes passed dbname which is used to do authentication in the proxy. Resulting connection string looks like that: `postgres://LizardWizzard%40zenith:<password>@start.stage.zenith.tech/old-sun-156649 dbname=postgres options='-c statement_timeout=120000'`

IMO the remote test shouldn't modify the passed connection string. One possible but hacky solution is to keep the parameters but raise an exception if any of them are set to something other than None and give back the pre-defined connection string. It is a hack but allows us to move forward here and do needed refactoring separately. Maybe since this pr is already a refactoring we can just change things here :)

Reply:
Ah of course. Ok, I'll bite the bullet and refactor this too.